docs/troubleshooting/vm.md

```diff
@@ -243,3 +243,174 @@ After the pod is successfully deleted, the `Start` button becomes visible again
+
+```
+time="2024-02-22T15:15:08.557487376Z" level=error msg="CreateContainer within sandbox \"d23af3219cb27228623cf8168ec27e64e836ed44f2b2f9cf784f0529a7f92e1e\" for &ContainerMetadata{Name:compute,Attempt:0,} failed" error="failed to generate container \"e4ed94fb5e9145e8716bcb87aae448300799f345197d52a617918d634d9ca3e1\" spec: failed to generate spec: get device path: /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/20291c6b-62c3-4456-be8a-fbeac118ec19 containerPath: /dev/disk-0 error: not a device node"
+```
+
+This is a CSI related file is an empty file instead of the expected device file. Then the containerd denied the `CreateContainer` request.
+
+```
+$ ls /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/ -alth
+total 8.0K
+drwxr-x--- 2 root root 4.0K Feb 22 15:10 .
+-rw-r--r-- 1 root root    0 Feb 22 14:28 aa851da3-cee1-45be-a585-26ae766c16ca
+-rw-r--r-- 1 root root    0 Feb 22 14:07 20291c6b-62c3-4456-be8a-fbeac118ec19
+drwxr-x--- 4 root root 4.0K Feb 22 14:06 ..
+-rw-r--r-- 1 root root    0 Feb 21 15:48 4333c9fd-c2c8-4da2-9b5a-1a310f80d9fd
+-rw-r--r-- 1 root root    0 Feb 21 09:18 becc0687-b6f5-433e-bfb7-756b00deb61b
+
+$file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-3c1b28fb-683e-4bf5-9869-c9107a0f1732/20291c6b-62c3-4456-be8a-fbeac118ec19
+: empty
+```
+
+By contrast, another running VM's related file is the expected device file.
+
+```
+$ ls  /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/publish/pvc-732f8496-103b-4a08-83af-8325e1c314b7/ -alth
+total 8.0K
+drwxr-x--- 2 root root  4.0K Feb 21 10:53 .
+drwxr-x--- 4 root root  4.0K Feb 21 10:53 ..
+brw-rw---- 1 root root 8, 16 Feb 21 10:53 4883af80-c202-4529-a2c6-4e7f15fe5a9b
+```
+
+#### Root Cause
+
+The kubelet calls `NodePublishVolume` for the new pod after node/cluster rebooting without first calling `NodeStageVolume`, and the Longhorn CSI plugin bind mounts the regular file at staging target path (previously used by the deleted pod) to target path and returns success.
+
+### Workaround
+
+Cluster level operation:
+
+1. Find the backing pods of the affected VMs and the related Longhorn volumes.
+
+    ```
+    $ kubectl get pods
+    NAME                      READY   STATUS                 RESTARTS   AGE
+    virt-launcher-vm1-nxfm4   0/2     CreateContainerError   0          7m11s
+
+    $ kubectl get pvc -A
+    NAMESPACE                  NAME                       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS           AGE
+    default                    vm1-disk-0-9gc6h           Bound    pvc-f1798969-5b72-4d76-9f0e-64854af7b59c   1Gi        RWX            longhorn-image-fxsqr   7d22h
+    ```
+
+1. [Stop](#vm-general-operations) the affected VMs from Harvester UI.
+
+   The VM may stuck in `Stopping`, continue the next step.
+
+1. Delete the backing pods forcely.
```

Comment:
NIT: There is no reason from the Longhorn side that the pods need to be deleted forcefully. Assuming the unmount step proceeds well, the pod that is stuck in `Terminating` will eventually be successfully cleaned up.

Reply:
I expect it might take up to five minutes, since kubelet is in exponential backoff trying to clean up the volumes. In my testing, the next time kubelet tried to unmap the volume after doing the unmount step, it was able to proceed and succeed.

Reply:
@ejweber 
Reproduced and retested in Harvester, the VM's POD is still stucking on `Terminating` more than 30 minutes later,  after `Stop` the VM on UI. A force deleting of the POD seems better.

```
$ while true; do kubectl get pods; date; sleep 10; done
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          23m
Mon Mar 11 14:44:28 UTC 2024
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          23m
Mon Mar 11 14:44:38 UTC 2024
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          23m
...
...
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          53m
Mon Mar 11 15:14:45 UTC 2024
NAME                      READY   STATUS        RESTARTS   AGE
virt-launcher-vm1-q2865   0/2     Terminating   0          53m
```

```
E0311 15:14:54.856070    2286 nestedpendingoperations.go:348] Operation for "{volumeName:kubernetes.io/csi/driver.longhorn.io^pvc-f1798969-5b72-4d76-9f0e-64854af7b59c podName:a676483b-ef5a-4eb4-969b-24038ce26448 nodeName:}" failed. No retries permitted until 2024-03-11 15:16:56.856056804 +0000 UTC m=+3511.059632246 (durationBeforeRetry 2m2s). Error: UnmapVolume.UnmapBlockVolume failed for volume "disk-0" (UniqueName: "kubernetes.io/csi/driver.longhorn.io^pvc-f1798969-5b72-4d76-9f0e-64854af7b59c") pod "a676483b-ef5a-4eb4-969b-24038ce26448" (UID: "a676483b-ef5a-4eb4-969b-24038ce26448") : blkUtil.DetachFileDevice failed. globalUnmapPath:/var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-f1798969-5b72-4d76-9f0e-64854af7b59c/dev, podUID: a676483b-ef5a-4eb4-969b-24038ce26448, bindMount: true: failed to remove file /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-f1798969-5b72-4d76-9f0e-64854af7b59c/dev/a676483b-ef5a-4eb4-969b-24038ce26448: remove /var/lib/kubelet/plugins/kubernetes.io/csi/volumeDevices/pvc-f1798969-5b72-4d76-9f0e-64854af7b59c/dev/a676483b-ef5a-4eb4-969b-24038ce26448: device or resource busy
```